torch/backends/_nnapi/serializer.py

```diff
@@ -0,0 +1,1131 @@
+        operand_id = len(self.operands)
+        self.operands.append(toper)
+        tsize = tensor_size(toper.op_type, toper.shape)
+        psize = ((tsize - 1) | 0x3) + 1
+        self.values.append((operand_id, OperandValueSourceType.NUMBERED_BUFFER))
+        if self.solid_weights:
+            buf_num = 0
+            offset = self.weight_offset
+            self.weight_offset += psize
+        else:
+            buf_num = len(self.used_weights)
+            offset = 0
+        self.value_data.append(struct.pack(
+            "iii",
+            buf_num,
+            offset,
+            tsize))
+        self.used_weights.append(tensor)
+        return operand_id
+
+
+    def add_immediate_operand(self, code, value, dims):
+        assert isinstance(dims, tuple)
+        cache_key = (code, value)
+        if cache_key not in self.cached_immediates:
+            operand_id = len(self.operands)
+            self.operands.append(Operand(code, dims, DimOrder.SCALAR_OR_VECTOR, 0.0, 0))
+            self.values.append((operand_id, OperandValueSourceType.IMMEDIATE))
+            self.value_data.append(value)
+            self.cached_immediates[cache_key] = operand_id
+        return self.cached_immediates[cache_key]
+
+
+    def add_immediate_int_scalar(self, value):
+        return self.add_immediate_operand(
+                NNAPI_OperandCode.INT32,
+                struct.pack("i", value),
+                ())
+
+
+    def add_immediate_float_scalar(self, value):
+        return self.add_immediate_operand(
+                NNAPI_OperandCode.FLOAT32,
+                struct.pack("f", value),
+                ())
+
+
+    def add_immediate_bool_scalar(self, value):
+        return self.add_immediate_operand(
+                NNAPI_OperandCode.BOOL,
+                b"\x01" if value else b"\x00",
+                ())
+
+
+    def get_tensor_operand_by_jitval(self, jitval):
+        operand_id = self.jitval_operand_map[jitval]
+        return (operand_id, self.operands[operand_id])
+
+
+    def get_tensor_operand_or_constant(self, jitval):
+        operand_id = self.jitval_operand_map.get(jitval)
+        if operand_id is None:
+            _, value = self.get_constant_value(jitval, "TensorType")
+            operand_id = self.add_tensor_operand_for_weight(value)
+        return (operand_id, self.operands[operand_id])
+
+
+    def get_tensor_operand_for_weight(self, jitval):
+        _, value = self.get_constant_value(jitval, "TensorType")
+        operand_id = self.add_tensor_operand_for_weight(value)
+        return (operand_id, self.operands[operand_id])
+
+
+    def add_operation(self, opcode, inputs, outputs):
+        self.operations.append((opcode, len(inputs), len(outputs)))
+        self.operation_args.extend(inputs + outputs)
+
+    def add_constant_value(self, jitval, ctype, value):
+        assert jitval not in self.constants
+        self.constants[jitval] = (ctype, value)
```

Comment:
Q. Where can I find out more about the permissible values in `jitval` and what they mean?

Reply:
This is a "Value" in the JIT IR.  https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/ir/ir.h#L148  It represents a local variable or the result of a TorchScript expression.

Reply:
I don't see a conversion from `struct Value` to `std::string` or `const char*` in the file `csrc/jit/ir/ir.h`. How does one use `jitval` as a string in this context?